### PR TITLE
Custom skincolors

### PR DESCRIPTION
**Pull request by SwitchKaze**
Allows Lua and SOC to create custom skincolors.
(clone of https://git.magicalgirl.moe/STJr/SRB2/merge_requests/783/ with merge conflicts resolved)

**skincolor_t**

| Name       | Type          | Access. | Description                                                                                                                         |
|------------|---------------|---------|-------------------------------------------------------------------------------------------------------------------------------------|
| name       | string        | R/W     | The name of the skincolor. Used in the console and Player Setup menu. Names containing spaces must be put in quotes in the console. |
| ramp       | `UINT8` array | R/W     | An array of 16 palette indices, which in whole represent the skincolor's color ramp.                                                |
| invcolor   | `UINT8`       | R/W     | The opposite skincolor. Used on the Goal Sign.                                                                                      |
| invshade   | `UINT8`       | R/W     | The shade of the opposite skincolor, from 0 to 15. Used on the Goal Sign.                                                           |
| chatcolor  | `UINT16`      | R/W     | Color to use when coloring a player name in the chat. Accepts `V_*MAP` constants.                                                     |
| accessible | boolean       | R/W     | Skincolor accessibility. Determines whether the color can be accessed from the Player Setup menu or from the console.               |

Both of these are identical to each other and functionally identical to the internal color `SKINCOLOR_BLUE`:
```Lua
freeslot("SKINCOLOR_BLUECLONE")
skincolors[SKINCOLOR_BLUECLONE] = {
	name = "BlueClone",
	ramp = {146,147,148,149,150,151,152,153,154,155,156,157,158,159,253,254},
	invcolor = SKINCOLOR_ORANGE,
	invshade = 9,
	chatcolor = V_BLUEMAP,
	accessible = true
}
```
```
FREESLOT
SKINCOLOR_BLUECLONE

SKINCOLOR SKINCOLOR_BLUECLONE
NAME = BlueClone
RAMP = 146,147,148,149,150,151,152,153,154,155,156,157,158,159,253,254
INVCOLOR = SKINCOLOR_ORANGE
INVSHADE = 9
CHATCOLOR = V_BLUEMAP
ACCESSIBLE = TRUE
```

Several functions have also been added. These functions can be used to get or set the order of skincolors as displayed in the Player Setup menu.

| Function                               | Returns | Description                                       |
|----------------------------------------|---------|---------------------------------------------------|
| M_MoveColorBefore(int color, int targ) | nil     | Moves the color `color` before the color `targ`.  |
| M_MoveColorAfter(int color, int targ)  | nil     | Moves the color  `color` after the color  `targ`. |
| M_GetColorBefore(int color)            | int     | Returns the color before the color `color`.       |
| M_GetColorAfter(int color)             | int     | Returns the color after the color  `color`.       |

Due to lump loading order being the way it is, a color must be defined in a Lua lump to be used in `S_SKIN`.
For a skin to use a custom super color `ColorName` in its `S_SKIN`, there must be exist 5 colors with names `Super ColorName 1` through `Super ColorName 5`.

Demonstration:

https://gfycat.com/slipperymadeupangelwingmussel

https://gfycat.com/drearyexcellentcrane

### CODE REVIEW

**Comment by alama** [comment]
Can you like, fix the build errors? try compiling SRB2 with `ERRORMODE=1` if you using the Makefile build system
you got:
- alot of `-Werror=type-limits` 
- 2 `-Werror=missing-prototypes` 
- even `old-style function definition [-Werror=old-style-definition]`

I mean who code in old ANSI C?

**Comment by SwitchKaze** [comment]
The build bots have asserted my worthiness.

I am honored.

**Comment by SwitchKaze** [comment]
Updated the code to resolve conflicts with 2.2.1.

**Comment by SwitchKaze** [comment]
Built-in colors can no longer be modified.

File: src/lua_baselib.c [hunk 1]
```diff
@@ -146,6 +147,8 @@ static const struct {
 	{META_STATE,        "state_t"},
 	{META_MOBJINFO,     "mobjinfo_t"},
 	{META_SFXINFO,      "sfxinfo_t"},
+	{META_SKINCOLOR,    "sfxinfo_t"},
```
[the remark this answers is not in the document]
Comment (SwitchKaze) [reply src/lua_baselib.c]: Yeah, I overlooked this

[the remark this answers is not in the document]
Comment (SwitchKaze) [reply src/lua_baselib.c]: Actually it looks like I fixed this in commit 58d43548?

File: src/p_enemy.c [hunk 2]
```diff
@@ -8831,7 +8820,41 @@ void A_ChangeColorAbsolute(mobj_t *actor)
...
+}
+
+// Function: A_Dye
+//
+// Description: Colorizes an object.
+//
+// var1 = if (var1 != 0), dye your target instead of yourself
+// var2 = color value to dye
+//
+void A_Dye(mobj_t *actor)
```
Comment (alama) [inline src/p_enemy.c:8833]:
```
p_enemy.c:8867:6: error: redefinition of 'A_Dye'
 void A_Dye(mobj_t *actor)
      ^~~~~
p_enemy.c:8833:6: note: previous definition of 'A_Dye' was here
 void A_Dye(mobj_t *actor)
      ^~~~~
p_enemy.c: In function 'A_Dye':
p_enemy.c:8876:15: error: 'MAXTRANSLATIONS' undeclared (first use in this function)
  if (color >= MAXTRANSLATIONS)
               ^~~~~~~~~~~~~~~
p_enemy.c:8876:15: note: each undeclared identifier is reported only once for each function it appears in
```

Comment (SwitchKaze) [reply src/p_enemy.c:8833]: I really should not have tried to merge the next-lc-uint16 branch into next-luacolors. I'll close this PR and make a new one with the new branch.

**Comment by SwitchKaze** [comment]
Closed in favor of #461.

**Pull request closed without merging**